module/zfs/dmu_objset.c

```diff
@@ -1639,20 +1640,67 @@ sync_dnodes_task(void *arg)
 
 	multilist_sublist_unlock(ms);
 
+	zio_nowait(sda->sda_cio);
+
 	kmem_free(sda, sizeof (*sda));
 }
 
+typedef struct sync_objset_arg {
+	objset_t *soa_os;
+	dmu_tx_t *soa_tx;
+	zio_t *soa_zio;
+	taskq_ent_t soa_tq_ent;
+} sync_objset_arg_t;
+
+static void
+sync_dnodes_finish_task(void *arg)
+{
+	sync_objset_arg_t *soa = arg;
+	objset_t *os = soa->soa_os;
+	dmu_tx_t *tx = soa->soa_tx;
+
+	/* Free intent log blocks up to this tx. (May block.) */
+	zil_sync(os->os_zil, tx);
+	os->os_phys->os_zil_header = os->os_zil_header;
+
+	zio_nowait(soa->soa_zio);
+	kmem_free(soa, sizeof (*soa));
+}
 
-/* called from dsl */
 void
-dmu_objset_sync(objset_t *os, zio_t *pio, dmu_tx_t *tx)
+dmu_objset_sync_sublists_done(zio_t *zio)
+{
+	sync_objset_arg_t *soa = zio->io_private;
+	objset_t *os = soa->soa_os;
+	dmu_tx_t *tx = soa->soa_tx;
+	int txgoff = tx->tx_txg & TXG_MASK;
+	dbuf_dirty_record_t *dr;
+
+	list_t *list = &DMU_META_DNODE(os)->dn_dirty_records[txgoff];
+	while ((dr = list_remove_head(list)) != NULL) {
+		ASSERT0(dr->dr_dbuf->db_level);
+		zio_nowait(dr->dr_zio);
+	}
+
+	/* Enable dnode backfill if enough objects have been freed. */
+	if (os->os_freed_dnodes >= dmu_rescan_dnode_threshold) {
+		os->os_rescan_dnodes = B_TRUE;
+		os->os_freed_dnodes = 0;
+	}
+
+	/* sync_dnodes_finsh_task calls zil_sync on our behalf. */
+	taskq_dispatch_ent(dmu_objset_pool(os)->dp_sync_taskq,
+	    sync_dnodes_finish_task, soa, TQ_FRONT, &soa->soa_tq_ent);
+}
+
+/* Nonblocking objset sync. Called from dsl. */
+zio_t *
```

Comment:
I'm not sure this signature change is necessary. There's only two call sites, and both immediately wait on the return, and formerly one of those set up a null zio so it could do a blocking wait anyway.

It possibly doesn't matter, but `dmu_objset_sync` is an exported symbol, and I'm always wary of changing the signature of any "public" symbol that I don't have to. So I wouldn't die on this hill but you might consider it.

I think this patch reverts the signature change while keeping the same behaviour:

```diff
diff --git include/sys/dmu_objset.h include/sys/dmu_objset.h
index 96bff4e4e..3c2ad73f3 100644
--- include/sys/dmu_objset.h
+++ include/sys/dmu_objset.h
@@ -235,7 +235,7 @@ void dmu_objset_evict_dbufs(objset_t *os);
 inode_timespec_t dmu_objset_snap_cmtime(objset_t *os);
 
 /* called from dsl */
-zio_t *dmu_objset_sync(objset_t *os, zio_t *zio, dmu_tx_t *tx);
+void dmu_objset_sync(objset_t *os, zio_t *zio, dmu_tx_t *tx);
 void dmu_objset_sync_sublists_done(zio_t *zio);
 boolean_t dmu_objset_is_dirty(objset_t *os, uint64_t txg);
 objset_t *dmu_objset_create_impl_dnstats(spa_t *spa, struct dsl_dataset *ds,
diff --git module/zfs/dmu_objset.c module/zfs/dmu_objset.c
index 20a334a00..409263112 100644
--- module/zfs/dmu_objset.c
+++ module/zfs/dmu_objset.c
@@ -1694,7 +1694,7 @@ dmu_objset_sync_sublists_done(zio_t *zio)
 }
 
 /* Nonblocking objset sync. Called from dsl. */
-zio_t *
+void
 dmu_objset_sync(objset_t *os, zio_t *rio, dmu_tx_t *tx)
 {
 	int txgoff;
@@ -1821,7 +1821,7 @@ dmu_objset_sync(objset_t *os, zio_t *rio, dmu_tx_t *tx)
 	/* Syncs the meta dnode even if all sublists were empty */
 	zio_nowait(sio);
 
-	return (pio);
+	zio_nowait(pio);
 }
 
 boolean_t
diff --git module/zfs/dsl_dataset.c module/zfs/dsl_dataset.c
index 0002e8264..b634d1713 100644
--- module/zfs/dsl_dataset.c
+++ module/zfs/dsl_dataset.c
@@ -2099,7 +2099,7 @@ dsl_dataset_sync(dsl_dataset_t *ds, zio_t *pio, dmu_tx_t *tx)
 		ds->ds_resume_bytes[tx->tx_txg & TXG_MASK] = 0;
 	}
 
-	zio_nowait(dmu_objset_sync(ds->ds_objset, pio, tx));
+	dmu_objset_sync(ds->ds_objset, pio, tx);
 }
 
 /*
diff --git module/zfs/dsl_pool.c module/zfs/dsl_pool.c
index 3ec374b1d..ef26775c2 100644
--- module/zfs/dsl_pool.c
+++ module/zfs/dsl_pool.c
@@ -573,7 +573,9 @@ dsl_pool_mos_diduse_space(dsl_pool_t *dp,
 static void
 dsl_pool_sync_mos(dsl_pool_t *dp, dmu_tx_t *tx)
 {
-	VERIFY0(zio_wait(dmu_objset_sync(dp->dp_meta_objset, NULL, tx)));
+	zio_t *zio = zio_root(dp->dp_spa, NULL, NULL, ZIO_FLAG_MUSTSUCCEED);
+	dmu_objset_sync(dp->dp_meta_objset, zio, tx);
+	VERIFY0(zio_wait(zio));
 	dmu_objset_sync_done(dp->dp_meta_objset, tx);
 	taskq_wait(dp->dp_sync_taskq);
 	multilist_destroy(&dp->dp_meta_objset->os_synced_dnodes);
```

Reply:
updated to restore previous useage, thanks.